lib/LanguageServerRename/Model/Renamer/MemberRenamer.php

```diff
@@ -0,0 +1,182 @@
+            || $node instanceof ClassConstDeclaration
+            || ($node instanceof MemberAccessExpression && $node->memberName instanceof Token)
+        ) {
+            return $node;
+        }
+        
+        return null;
+    }
+
+    public function prepareRename(TextDocument $textDocument, ByteOffset $offset): ?ByteOffsetRange
+    {
+        if (($node = $this->getValidNode($textDocument, $offset)) !== null) {
+            [ $token ] = $this->getNodeNameTokens($node);
+            return $this->nodeUtils->getTokenNameRange($token, (string)$textDocument);
+        }
+        
+        return null;
+    }
+    /**
+     * {@inheritDoc}
+     */
+    public function rename(TextDocument $textDocument, ByteOffset $offset, string $newName): Generator
+    {
+        if (($node = $this->getValidNode($textDocument, $offset)) === null) {
+            return null;
+        }
+
+        [ $token ] = $this->getNodeNameTokens($node);
+        $oldName = $this->nodeUtils->getTokenNameText($token, (string)$textDocument);
+
+        foreach ($this->locationsProvider->provideLocations($textDocument, $offset) as $locationGroup) {
+            /** @var RenameLocationGroup $locationGroup */
+            yield new RenameResult(
+                $this->locationsToTextEdits($locationGroup->uri(), $locationGroup->locations(), $oldName, $newName),
+                $locationGroup->uri()
+            );
+        }
+    }
+    /** @param Location[] $locations */
+    private function locationsToTextEdits(TextDocumentUri $textDocumentUri, array $locations, string $oldName, string $newName): TextEdits
+    {
+        $textEdits = [];
+        $rootNode = $this->parser->parseSourceFile($this->locator->get($textDocumentUri));
+        foreach ($locations as $location) {
+            $node = $rootNode->getDescendantNodeAtPosition($location->offset()->toInt());
+            $tokens = $this->getNodeNameTokens($node);
+
+            foreach ($tokens as $token) {
+                $range = $this->nodeUtils->getTokenNameRange($token, $rootNode->getFileContents());
+                $rangeText = $this->nodeUtils->getTokenNameText($token, $rootNode->getFileContents());
+
+                if ($rangeText == $oldName) {
+                    $textEdits[] = TextEdit::create($range->start(), $range->end()->toInt() - $range->start()->toInt(), $newName);
+                    break;
+                }
+            }
+        }
+        return TextEdits::fromTextEdits($textEdits);
+    }
+    /** @return Token[] */
+    public function getNodeNameTokens(Node $node): array
+    {
+        if ($node instanceof MethodDeclaration && $node->name !== null) {
+            return [ $node->name ];
+        }
+
+        if ($node instanceof ScopedPropertyAccessExpression && $node->memberName instanceof Token) {
+            return [ $node->memberName ];
+        }
+        
+        if ($node instanceof Variable && $node->name instanceof Token) {
+            return [ $node->name ];
```

Comment:
what is this for?

Reply:
When you are standing like this:
```php
property $pro<>p1;
```
The underlying token is a `Variable`. The parents are validated in `getValidToken` - `ScopedPropertyAccessExpression` and `PropertyDeclaration`.